From d81a658c13631213fe74081cef62fd3985d1a788 Mon Sep 17 00:00:00 2001
From: Pushpal Sidhu <psidhu@gateworks.com>
Date: Fri, 1 Jul 2016 09:45:52 -0700
Subject: [PATCH 1/4] src: mtd: fix compilation with certain c libraries

Signed-off-by: Tim Harvey <tharvey@gateworks.com>
---
 src/mtd.c | 1 +
 src/mtd.h | 2 ++
 2 files changed, 3 insertions(+)

diff --git a/src/mtd.c b/src/mtd.c
index 9478756..beeb153 100644
--- a/src/mtd.c
+++ b/src/mtd.c
@@ -28,6 +28,7 @@
 #include <unistd.h>
 #include <stdlib.h>
 #include <string.h>
+#include <stddef.h>
 #include <fcntl.h>
 #include <ctype.h>
 #include <errno.h>
diff --git a/src/mtd.h b/src/mtd.h
index f5504ed..57852da 100644
--- a/src/mtd.h
+++ b/src/mtd.h
@@ -25,8 +25,10 @@
 #ifndef MTD_H
 #define MTD_H
 
+#define _GNU_SOURCE
 #include <mtd/mtd-user.h>
 #include <endian.h>
+#include <fcntl.h>
 
 #include "BootControlBlocks.h"
 #include "rom_nand_hamming_code_ecc.h"
-- 
2.8.3

