diff --git a/drivers/net/wireless/ath/ath10k/thermal.c b/drivers/net/wireless/ath/ath10k/thermal.c
index b14ae8d..e3f136a 100644
--- a/drivers/net/wireless/ath/ath10k/thermal.c
+++ b/drivers/net/wireless/ath/ath10k/thermal.c
@@ -92,7 +92,6 @@ static int ath10k_thermal_set_cur_dutycycle(struct thermal_cooling_device *cdev,
 	 */
 	num_bss = ath10k_thermal_get_active_vifs(ar, WMI_VDEV_TYPE_AP);
 	if (!num_bss) {
-		ath10k_warn(ar, "no active AP interfaces\n");
 		ret = -ENETDOWN;
 		goto out;
 	}
